plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -214,40 +216,108 @@ private void destroy()
     {
         private static final int INSTANCE_SIZE = ClassLayout.parseClass(BufferedStorageWriter.class).instanceSize();
 
-        private final ExchangeStorageWriter storageWriter;
+        private final FileSystemExchangeStorage exchangeStorage;
+        private final URI outputDirectory;
+        private final Optional<SecretKey> secretKey;
+        private final int partitionId;
         private final BufferPool bufferPool;
         private final AtomicReference<Throwable> failure;
+        private final int maxPageStorageSizeInBytes;
+        private final long maxFileSizeInBytes;
 
+        @GuardedBy("this")
+        private ExchangeStorageWriter currentWriter;
+        @GuardedBy("this")
+        private long currentFileSize;
         @GuardedBy("this")
         private SliceOutput currentBuffer;
+        @GuardedBy("this")
+        private final List<ExchangeStorageWriter> writers = new ArrayList<>();
+        @GuardedBy("this")
+        private boolean closed;
 
-        public BufferedStorageWriter(ExchangeStorageWriter storageWriter, BufferPool bufferPool, AtomicReference<Throwable> failure)
+        public BufferedStorageWriter(
+                FileSystemExchangeStorage exchangeStorage,
+                URI outputDirectory,
+                Optional<SecretKey> secretKey,
+                int partitionId,
+                BufferPool bufferPool,
+                AtomicReference<Throwable> failure,
+                int maxPageStorageSizeInBytes,
+                long maxFileSizeInBytes)
         {
-            this.storageWriter = requireNonNull(storageWriter, "storageWriter is null");
+            this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+            this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+            this.partitionId = partitionId;
             this.bufferPool = requireNonNull(bufferPool, "bufferPool is null");
             this.failure = requireNonNull(failure, "failure is null");
+            this.maxPageStorageSizeInBytes = maxPageStorageSizeInBytes;
+            this.maxFileSizeInBytes = maxFileSizeInBytes;
+
+            updateCurrentWriter();
         }
 
         public synchronized void write(Slice data)
         {
+            if (closed) {
+                return;
+            }
+
+            int requiredPageStorageSize = data.length() + Integer.BYTES;
+            if (requiredPageStorageSize > maxPageStorageSizeInBytes) {
+                throw new TrinoException(NOT_SUPPORTED, format("Max row size of %s exceeded: %s", succinctBytes(maxPageStorageSizeInBytes), succinctBytes(requiredPageStorageSize)));
+            }
+
+            if (currentFileSize + requiredPageStorageSize > maxFileSizeInBytes) {
```

Comment:
Totally forgot. When `preserveRecordsOrder` is requested only a single file must be created and this optimization must be disabled.